File: pt5/1104.path-in-zigzag-labelled-binary-tree.py
from __future__ import annotations

from typing import List


class Solution:
    def pathInZigZagTree(self, label: int) -> List[int]:
        path = []

        while label:
            path.append(label)
            level_start = 1 << (label.bit_length() - 1)
            level_end = (level_start << 1) - 1
            label = (level_start + level_end - label) // 2

        return path[::-1]

File: pt5/1104.path-in-zigzag-labelled-binary-tree.java
import java.util.*;

class Solution {
    public List<Integer> pathInZigZagTree(int label) {
        List<Integer> path = new ArrayList<>();

        while (label > 0) {
            path.add(label);
            int levelStart = Integer.highestOneBit(label);
            int levelEnd = (levelStart << 1) - 1;
            label = (levelStart + levelEnd - label) / 2;
        }

        Collections.reverse(path);
        return path;
    }
}

